src/style-spec/function/definitions/match.js

```diff
@@ -0,0 +1,121 @@
+// @flow
+
+const assert = require('assert');
+const { parseExpression, match } = require('../expression');
+const { typeOf } = require('../values');
+
+import type { Expression, ParsingContext } from '../expression';
+import type { Type } from '../types';
+
+type Branches = Array<[Array<null | number | string | boolean>, Expression]>;
+
+class Match implements Expression {
+    key: string;
+    type: Type;
+    inputType: Type;
+
+    input: Expression;
+    branches: Branches;
+    otherwise: Expression;
+
+    constructor(key: string, inputType: Type, outputType: Type, input: Expression, branches: Branches, otherwise: Expression) {
+        this.key = key;
+        this.inputType = inputType;
+        this.type = outputType;
+        this.input = input;
+        this.branches = branches;
+        this.otherwise = otherwise;
+    }
+
+    static parse(args: Array<mixed>, context: ParsingContext) {
+        args = args.slice(1);
+        if (args.length < 2)
+            return context.error(`Expected at least 2 arguments, but found only ${args.length}.`);
+        if (args.length % 2 !== 0)
+            return context.error(`Expected an even number of arguments.`);
+
+        let inputType;
+        let outputType = context.expectedType;
+        const branches = [];
+        for (let i = 1; i < args.length - 1; i += 2) {
+            let labels = args[i];
+            const value = args[i + 1];
+
+            if (!Array.isArray(labels)) {
+                labels = [labels];
+            }
+
+            const labelContext = context.concat(i + 1);
+            if (labels.length === 0) {
+                return labelContext.error('Expected at least one branch label.');
+            }
+
+            for (const label of labels) {
+                if (label !== null && typeof label !== 'number' && typeof label !== 'string' && typeof label !== 'boolean') {
+                    return labelContext.error(`Branch labels must be null, numbers, strings, or booleans.`);
+                } else if (!inputType) {
+                    inputType = typeOf(label);
+                } else if (match(inputType, typeOf(label), labelContext)) {
+                    return null;
+                }
+            }
+
+            const result = parseExpression(value, context.concat(i + 1, outputType));
+            if (!result) return null;
+            outputType = outputType || result.type;
+
+            branches.push([(labels: any), result]);
+        }
+
+        const input = parseExpression(args[0], context.concat(1, inputType));
+        if (!input) return null;
+
+        const otherwise = parseExpression(args[args.length - 1], context.concat(args.length, outputType));
+        if (!otherwise) return null;
+
+        assert(inputType && outputType);
```

Comment:
`match` with only two arguments triggers this assertion -- `inputType` is undefined.